### PR TITLE
Add convolution Function

### PR DESCRIPTION
**Pull request by hedaoyuan**
Follow comments in issue #2196 
* [x] Defines the data structure of the input, output, and filter.
* [x] Migrate the code based on sgemm to the `GemmConvOp.cpp`.
* [x] Refactor the ExpandConvBaseLayer, ExpandConvLayer and ExpandConvTransLayer.

Code refactoring:

| number | old code | new code | about |
|----|----|----|----|
| 1 | ExpandConvTransLayer, ExpandConvLayer | ExpandConvLayer | Remove ExpandConvTransLayer, both exconv and exconvt are based on ExpandConvLayer. |
| 2 | ExpandConvBaseLayer | ConvFunctionBase | |
| 3 | ExpandConvBaseLayer::expandOneFrame, Matrix::convExpand | Im2ColFunctor | Refactor code. |
| 4 | Matrix::convShrink | Col2ImFunctor | Refactor code. |
| 5 | ExpandConvBaseLayer::expandFwdOnce | GemmConvFunction | Refactor code. |
| 6 | ExpandConvBaseLayer::bpropActs | GemmConvGradInputFunction | Refactor code. |
| 7 | ExpandConvBaseLayer::bpropWeights | GemmConvGradFilterFunction | Refactor code. |
| 8 | | NaiveConvFunction | Add a naive convolution calculation. |
| 9 | FunctionCompare | Compare2Function, CpuGpuFuncCompare | Replace FunctionCompare with the Compare2Function type. The original FunctionCompare is actually a CpuGpuFuncCompare. |
| 10 | | ConvolutionTest | Added unit tests for the implementation of various convolution calculations. |

--------------
To do:
1. Remove and Matrix::convExpand and Matrix::convShrink and Refactor BlockExpandLayer #2424 
2. Output size calculation #2460 .
3. Migrate the code based on cudnn to the `CudnnConvOp.cpp`. #2425

### CODE REVIEW

File: paddle/function/ConvOpTest.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+public:
+  ConvolutionTest(const std::string& conv1,
+                  const std::string& conv2,
+                  TestType type,
+                  std::string algo = "auto") {
+    for (size_t batchSize : {1, 32}) {
+      for (size_t inputSize : {7, 14, 54}) {
+        for (size_t filterSize : {1, 3, 5}) {
+          for (size_t inputChannels : {3, 64}) {
+            for (size_t outputChannels : {3, 64, 128}) {
```
Comment (qingqing01) [inline paddle/function/ConvOpTest.cpp]: 是否需要增加长方形input,output, filter的单测？

File: paddle/function/GemmConvOp.cpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+                  int inputWidth,
+                  int filterHeight,
+                  int filterWidth,
+                  int strideHeight,
+                  int strideWidth,
+                  int paddingHeight,
+                  int paddingWidth,
+                  int outputHeight,
+                  int outputWidth,
+                  T* colData) {
```
Comment (qingqing01) [inline paddle/function/GemmConvOp.cpp]: 感觉接口直接传TensorShape会简单一些吧，就不用这么多参数了~

File: paddle/function/ConvOpTest.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#include "Function.h"
+#include "FunctionTest.h"
+
+namespace paddle {
+
+enum TestType {
+  FORWARD_TEST = 0,
+  BACKWARD_INPUT_TEST = 1,
+  BACKWARD_FILTER_TEST = 2,
+};
```
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: Done.

File: paddle/function/ConvOpTest.cpp (hunk 1 above)
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: 在这里加会让TEST变的时间很长，后续我单独增加针对长方形的测试吧。

File: paddle/function/ConvOpTest.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        for (size_t filterSize : {1, 3, 5}) {
+          for (size_t inputChannels : {3, 64}) {
+            for (size_t outputChannels : {3, 64, 128}) {
+              if (inputChannels < outputChannels) break;
+              for (size_t stride : {1, 2}) {
+                for (size_t padding : {0, 1}) {
+                  if (padding >= filterSize) break;
+                  size_t outputSize =
+                      (inputSize - filterSize + 2 * padding + stride) / stride;
+                  LOG(INFO) << " batchSize=" << batchSize
```
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: Done.

File: paddle/function/GemmConvOp.cpp (hunk 2 above)
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 修改成TensorShape的话，接口里面还需要做一次input/filter shape对应参数的检查。而且，这里没有batchSize这个维度，还不能直接用Function里面传入的参数的shape。

File: paddle/function/ConvOpTest.cpp (hunk 1 above)
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: Done.

File: paddle/function/GemmConvOp.cpp (hunk 2 above)
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: Will be fixed in the next [PR 2449](
https://github.com/PaddlePaddle/Paddle/pull/2449/files#diff-cdb1a42982348fbcf0f5b69ace9a7bdcR47).
```
class Im2ColFunctor {
public:
  void operator()(float* colData,
                  const float* imData,
                  const TensorShape& colShape,
                  const TensorShape& imShape,
                  int strideHeight,
                  int strideWidth,
                  int paddingHeight,
                  int paddingWidth);
};
```

File: paddle/gserver/layers/ExpandConvLayer.cpp [hunk 5]
```diff
@@ -18,32 +18,94 @@ limitations under the License. */
...
+
+  size_t numInputs = config_.inputs_size();
+  inputShape_.resize(numInputs);
+  filterShape_.resize(numInputs);
+  outputShape_.resize(numInputs);
+  for (int i = 0; i < config_.inputs_size(); i++) {
+    std::vector<size_t> paddings = {(size_t)paddingY_[i], (size_t)padding_[i]};
+    std::vector<size_t> strides = {(size_t)strideY_[i], (size_t)stride_[i]};
+    createFunction(forward_,
+                   !isDeconv_ ? "GemmConv" : "GemmConvGradInput",
```
Comment (qingqing01) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: isDeconv_ -> isConvTrans_

File: paddle/gserver/layers/ExpandConvLayer.cpp [hunk 6]
```diff
@@ -18,32 +18,94 @@ limitations under the License. */
...
+
+  // Calculate the output value.
+  for (size_t i = 0; i < inputLayers_.size(); ++i) {
+    BufferArgs inputs;
+    BufferArgs outputs;
+    inputs.addArg(*getInputValue(i), inputShape_[i]);
+    inputs.addArg(*weights_[i]->getW(), filterShape_[i]);
+    outputs.addArg(*getOutputValue(),
+                   outputShape_[i],
+                   !isDeconv_ && i == 0 ? ASSIGN_TO : ADD_TO);
```
Comment (qingqing01) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: isDeconv_为什么需要这个参数？

File: paddle/function/ConvOp.h [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    } else {
+      return filter[2];
+    }
+  }
+
+  size_t getFilterWidth(const TensorShape& filter) const {
+    if (filter.ndims() == 5) {
+      return filter[4];
+    } else {
+      return filter[3];
```
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: 没有理解这个comment，filter[-1],filter[-2]指的是啥？

File: paddle/function/ConvOpTest.cpp [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+                    test.run();
+                  } else if (type == kBackwardInputTest) {
+                    test.addInputs(BufferArg(VALUE_TYPE_FLOAT, output));
+                    test.addInputs(BufferArg(VALUE_TYPE_FLOAT, filter));
+                    test.addOutputs(BufferArg(VALUE_TYPE_FLOAT, input), ADD_TO);
+                    test.run();
+                  } else if (type == kBackwardFilterTest) {
+                    test.addInputs(BufferArg(VALUE_TYPE_FLOAT, output));
+                    test.addInputs(BufferArg(VALUE_TYPE_FLOAT, input));
+                    test.addOutputs(BufferArg(VALUE_TYPE_FLOAT, filter));
```
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: BackwardFilter是支持ASSIGN_TO模式的。Test可以默认都用ASSIGN_TO测试，对于不支持ASSIGN_TO的Function，需要显示配置ADD_TO。

File: paddle/function/FunctionTest.h [hunk 9]
```diff
@@ -236,27 +289,27 @@ class FunctionCompare {
...
 
     // check outputs
     compareOutputs();
   }
 
-  std::shared_ptr<FunctionBase> getCpuFunction() const { return cpuFunc_; }
+  std::shared_ptr<FunctionBase> getCpuFunction() const { return function1_; }
 
-  std::shared_ptr<FunctionBase> getGpuFunction() const { return gpuFunc_; }
+  std::shared_ptr<FunctionBase> getGpuFunction() const { return function2_; }
```
Comment (hedaoyuan) [inline paddle/function/FunctionTest.h]: 嗯，这里接口名称需要修改。

File: paddle/function/GemmConvOp.cpp [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+
+namespace paddle {
+
+/*
+ * imData = [input_channels, input_height, input_width]
+ * colData = [input_channels, filter_height, filter_width,
+ *            output_height, output_width]
+ */
+template <class T>
+class Im2ColFunctor<DEVICE_TYPE_CPU, T> {
```
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 在PR #2424 里面会移到一个单独的文件里面，但不是Function，被ImageExpandFunction调用。

File: paddle/function/GemmConvOp.cpp [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+
+    size_t batchSize = input[0];
+    size_t inputChannels = input[1];
+    size_t inputHeight = input[2];
+    size_t inputWidth = input[3];
+    size_t filterHeight = getFilterHeight(filter);
+    size_t filterWidth = getFilterWidth(filter);
+    size_t outputChannels = output[1];
+    size_t outputHeight = output[2];
+    size_t outputWidth = output[3];
```
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 个人觉得Function的代码逻辑可读性比较重要。如果用`n, ci, hi...`这种缩写的话，对于不熟悉代码逻辑的人并不一定很友好。

File: paddle/gserver/layers/ExpandConvLayer.cpp (hunk 5 above)
Comment (hedaoyuan) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: isDeconv_是`ConvBaseLayer`里面原来的名字，换成`isConvTrans_`更好？

File: paddle/gserver/layers/ExpandConvLayer.cpp (hunk 6 above)
Comment (hedaoyuan) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: 主要是目前GemmConvGradInput的实现只支持ADD_TO模式。
https://github.com/PaddlePaddle/Paddle/pull/2282/files/2608c4854273e47bd0958fbb03ca67050ddfb35c#diff-b71c34e8a73d71627e2b185b4dd33aafR215

File: paddle/function/ConvOp.h [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+ * Arguments format of input, filter and output:
+ *   1. Input image, output image, input image gradient, output image gradient
+ *      are all NCHW format. Where N is batch size, C is the number of channels,
+ *      H and W is the height and width of image or image gradient.
+ *
+ *   2. The format of the filter data is MCHW, where M is the number of output
+ *      image channels, C is the number of input image channels,
+ *      H and W is height and width of filter.
+ *
+ *      If groups is greater than 1, the filter's data format should be GMCHW,
```
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: 这个注释里面我觉得不用修改成groups_，FuncConfig的配置参数名称上也是groups。

File: paddle/function/ConvOp.h (hunk 7 above)
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: -1, -2指的是倒数第一个第二个值这样？

File: paddle/function/ConvOp.h [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+      // The input and output channel dimensions are the third and second
+      // dimensions of the filter shape.
+      CHECK_EQ(input[1], filter[2] * groups_);
+      CHECK_EQ(output[1], filter[1] * groups_);
+    }
+  }
+
+protected:
+  size_t getFilterHeight(const TensorShape& filter) const {
+    if (filter.ndims() == 5) {
```
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: Done.

Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: Done.

File: paddle/function/ConvOp.h [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+      return filter[3];
+    } else {
+      return filter[2];
+    }
+  }
+
+  size_t getFilterWidth(const TensorShape& filter) const {
+    if (filter.ndims() == 5) {
```
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: Done.

File: paddle/function/ConvOp.h (hunk 7 above)
Comment (hedaoyuan) [inline paddle/function/ConvOp.h]: Done.

File: paddle/function/ConvOp.h (hunk 12 above)
Comment (qingqing01) [inline paddle/function/ConvOp.h]: 上面提出修改原因是：`If groups is greater than 1` 不符合语法，也提醒我以后comments需要解释清楚 :)

File: paddle/function/GemmConvOpGpu.cu [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+      int h_col_start =
+        (h < (int)blockH) ? 0 : (h - (int)blockH) / (int)strideH + 1;
+      int h_col_end = min(int(h / strideH + 1), int(height_col));
+      for (int h_col = h_col_start; h_col < h_col_end; ++h_col) {
+        for (int w_col = w_col_start; w_col < w_col_end; ++w_col) {
+          // the col location: [c * width * height + h_out, w_out]
+          int c_col = int(c * blockH* blockW) + \
+            (h - h_col * (int)strideH) * (int)blockW +
+            (w - w_col * (int)strideW);
+          val += data_col[(c_col * height_col + h_col) * width_col + w_col];
```
Comment (qingqing01) [inline paddle/function/GemmConvOpGpu.cu]: cuda kernel里有多种命名规则~

File: paddle/gserver/layers/ExpandConvLayer.cpp (hunk 6 above)
Comment (qingqing01) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: 明白了。感觉Function是不是应该和blas库类似，加scale/beta这样的系数更合理些，而不是`ASSIGN_TO`、`ADD_TO`~

File: paddle/gserver/layers/ExpandConvLayer.cpp (hunk 5 above)
Comment (qingqing01) [inline paddle/gserver/layers/ExpandConvLayer.cpp]: 嗯，命名和对用户接口一致叫做convolution transpose layer觉得更好些，这样也可以吧~

**Pull request merged**